<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="RNeXML">
<title>Extending NeXML: an example based on simmap • RNeXML</title>
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="/apple-touch-icon.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://docs.ropensci.org/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Extending NeXML: an example based on simmap">
<meta property="og:description" content="RNeXML">
<meta property="og:image" content="https://docs.ropensci.org/RNeXML/logo.svg">
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Matomo --><link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/cookieconsent@3/build/cookieconsent.min.css">
<script src="https://cdn.jsdelivr.net/npm/cookieconsent@3/build/cookieconsent.min.js" data-cfasync="false"></script><script src="https://ropensci.org/scripts/matomo.js"></script><noscript><p><img src="https://ropensci.matomo.cloud/matomo.php?idsite=1&amp;rec=1" style="border:0;" alt=""></p></noscript>
<!-- End Matomo Code -->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    <a href="https://ropensci.org" class="external-link"><img src="https://ropensci.org/img/icon_short_white.svg" id="hexlogo" alt="rOpenSci"></a>
    <a class="navbar-brand me-2" href="../index.html">RNeXML</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">2.4.11</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"></ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/S4.html">The nexml S4 Object</a>
    <a class="dropdown-item" href="../articles/intro.html">A Brief Introduction to RNeXML</a>
    <a class="dropdown-item" href="../articles/metadata.html">Handling Metadata in RNeXML</a>
    <a class="dropdown-item" href="../articles/simmap.html">Extending NeXML: an example based on simmap</a>
    <a class="dropdown-item" href="../articles/sparql.html">SPARQL with RNeXML</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ropensci/RNeXML/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Extending NeXML: an example based on simmap</h1>
                        <h4 data-toc-skip class="author">Carl Boettiger</h4>
                        <h4 data-toc-skip class="author">Scott Chamberlain</h4>
                        <h4 data-toc-skip class="author">Rutger Vos</h4>
                        <h4 data-toc-skip class="author">Hilmar Lapp</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ropensci/RNeXML/blob/HEAD/vignettes/simmap.Rmd" class="external-link"><code>vignettes/simmap.Rmd</code></a></small>
      <div class="d-none name"><code>simmap.Rmd</code></div>
    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Extending NeXML to simmap}
-->
<div class="section level2">
<h2 id="extending-the-nexml-standard-through-metadata-annotation-">Extending the NeXML standard through metadata annotation.<a class="anchor" aria-label="anchor" href="#extending-the-nexml-standard-through-metadata-annotation-"></a>
</h2>
<p>Here we illustrate this process using the example of stochastic character mapping <span class="citation">(Huelsenbeck, Nielsen, and Bollback 2003)</span>. A stochastic character map is simply an annotation of the branches on a phylogeny, assigning each section of each branch to a particular “state” (typically of a morphological characteristic).</p>
<p><span class="citation">Bollback (2006)</span> provides a widely used stand-alone software implementation of this method in the software <code>simmap</code>, which modified the standard Newick tree format to express this additional information. This can break compatibility with other software, and creates a format that cannot be interpreted without additional information describing this convention. By contrast, the NeXML extension is not only backwards compatible but contains a precise and machine-readable description of what it is encoding.</p>
<p>In this example, we illustrate how the additional information required to define a stochastic character mapping (a <code>simmap</code> mapping) in NeXML.</p>
<p><span class="citation">Revell (2012)</span> describes the <code>phytools</code> package for R, which includes utilities for reading, manipulating, and writing <code>simmap</code> files in R. In this example, we also show how to define <code>RNeXML</code> functions that map the R representation used by Revell (an extension of the <code>ape</code> class) into the NeXML extension we have defined by using <code>RNeXML</code> functions.</p>
<p>Since a stochastic character map simply assigns different states to parts of a branch (or edge) on the phylogenetic tree, we can create a NeXML representation by annotating the <code>edge</code> elements with appropriate <code>meta</code> elements. These elements need to describe the character state being assigned and the duration (in terms of branch-length) that the edge spends in that state (Stochastic character maps are specific to time-calibrated or ultrametric trees).</p>
<p>NeXML already defines the <code>characters</code> element to handle discrete character traits (<code>nex:char</code>) and the states they can assume (<code>nex:state</code>). We will thus reuse the <code>characters</code> element for this purpose, referring to both the character trait and the states by the ids assigned to them in that element. (NeXML’s convention of referring to everything by id permits a single canonical definition of each term, making it clear where additional annotation belongs). For each edge, we need to indicate:</p>
<ul>
<li>That our annotation contains a stochastic character mapping reconstruction</li>
<li>Since many reconstructions are possible for a single edge, we give each reconstruction an id</li>
<li>We indicate for which character trait we are defining the reconstruction</li>
<li>We then indicate which states the character assumes on that edge. For each state realized on the edge, that involves stating:
<ul>
<li>the state assignment</li>
<li>the duration (length of time) for which the edge spends in the given state</li>
<li>the order in which the state changes happen (Though we could just assume state transitions are listed chronologically, NeXML suggests making all data explicit, rather than relying on the structure of the data file to convey information).</li>
</ul>
</li>
</ul>
<p>Thus the annotation for an edge that switches from state <code>s2</code> to state <code>s1</code> of character <code>cr1</code> would be constructed like this:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/meta.html">meta</a></span><span class="op">(</span><span class="st">"simmap:reconstructions"</span>, children <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="../reference/meta.html">meta</a></span><span class="op">(</span><span class="st">"simmap:reconstruction"</span>, children <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span></span>
<span>          <span class="fu"><a href="../reference/meta.html">meta</a></span><span class="op">(</span><span class="st">"simmap:char"</span>, <span class="st">"cr1"</span><span class="op">)</span>,</span>
<span>          <span class="fu"><a href="../reference/meta.html">meta</a></span><span class="op">(</span><span class="st">"simmap:stateChange"</span>, children <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>            <span class="fu"><a href="../reference/meta.html">meta</a></span><span class="op">(</span><span class="st">"simmap:order"</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="../reference/meta.html">meta</a></span><span class="op">(</span><span class="st">"simmap:length"</span>, <span class="st">"0.2030"</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="../reference/meta.html">meta</a></span><span class="op">(</span><span class="st">"simmap:state"</span>, <span class="st">"s2"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>          </span>
<span>          <span class="fu"><a href="../reference/meta.html">meta</a></span><span class="op">(</span><span class="st">"simmap:char"</span>, <span class="st">"cr1"</span><span class="op">)</span>,</span>
<span>          <span class="fu"><a href="../reference/meta.html">meta</a></span><span class="op">(</span><span class="st">"simmap:stateChange"</span>, children <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>            <span class="fu"><a href="../reference/meta.html">meta</a></span><span class="op">(</span><span class="st">"simmap:order"</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="../reference/meta.html">meta</a></span><span class="op">(</span><span class="st">"simmap:length"</span>, <span class="st">"0.0022"</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="../reference/meta.html">meta</a></span><span class="op">(</span><span class="st">"simmap:state"</span>, <span class="st">"s1"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>          <span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Of course writing out such a definition manually becomes tedious quickly. Because these are just R commands, we can easily define a function that can loop over an assignment like this for each edge, extracting the appropriate order, length and state from an existing R object such as that provided in the <code>phytools</code> package.<br>
Likewise, it is straightforward to define a function that reads this data using the <code>RNeXML</code> utilities and converts it back to the <code>phytools</code> package. The full implementation of this mapping can be seen in the <code><a href="../reference/simmap_to_nexml.html">simmap_to_nexml()</a></code> and the <code><a href="../reference/simmap_to_nexml.html">nexml_to_simmap()</a></code> functions provided in the <code>RNeXML</code> package.</p>
<p>As the code indicates, the key step is simply to define the data in meta elements. In so doing, we have defined a custom namespace, <code>simmap</code>, to hold our variables. This allows us to provide a URL with more detailed descriptions of what each of these elements mean:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_namespaces.html">add_namespaces</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>simmap <span class="op">=</span> <span class="st">"https://github.com/ropensci/RNeXML/tree/master/inst/simmap.md"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>At that URL we have posted a simple description of each term.</p>
<p>Using this convention we can generate NeXML files containing <code>simmap</code> data, read those files into R, and convert them back into the <code>phytools</code> package format. These simple functions serve as further illustration of how <code>RNeXML</code> can be used to extend the NeXML standard. We illustrate their use briefly here, starting with loading a <code>nexml</code> object containing a <code>simmap</code> reconstruction into R:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples"</span>, <span class="st">"simmap_ex.xml"</span>, package <span class="op">=</span> <span class="st">"RNeXML"</span><span class="op">)</span></span>
<span><span class="va">simmap_ex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/nexml_read.html">read.nexml</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/get_trees.html">get_trees()</a></code> function can be used to return an <code><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">ape::phylo</a></code> tree as usual. <code>RNeXML</code> automatically detects the <code>simmap</code> reconstruction data and returns includes this in a <code>maps</code> element of the <code><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">ape::phylo</a></code> object, for use with other <code>phytools</code> functions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simmap_to_nexml.html">nexml_to_simmap</a></span><span class="op">(</span><span class="va">simmap_ex</span><span class="op">)</span></span></code></pre></div>
<p>We can then use various functions from <code>phytools</code> designed for <code>simmap</code> objects <span class="citation">(Revell 2012)</span>, such as the plotting function:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/liamrevell/phytools" class="external-link">"phytools"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/plotSimmap.html" class="external-link">plotSimmap</a></span><span class="op">(</span><span class="va">phy</span><span class="op">)</span></span></code></pre></div>
<pre><code>no colors provided. using the following legend:
       A        B        C 
 "black"    "red" "green3" </code></pre>
<div class="figure">
<img src="simmap-Figure1-1.png" alt=""><p class="caption">Stochastic character mapping on a phylogeny, as generated by the phytools package after parsing the simmap-extended NeXML.</p>
</div>
<p>Likewise, we can convert the object back in the NeXML format and write it out to file to be read by other users.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simmap_to_nexml.html">simmap_to_nexml</a></span><span class="op">(</span><span class="va">phy</span><span class="op">)</span> </span>
<span><span class="fu"><a href="../reference/nexml_write.html">nexml_write</a></span><span class="op">(</span><span class="va">nex</span>, <span class="st">"simmap.xml"</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] "simmap.xml"</code></pre>
<p>Though other NeXML parsers (for instance, for Perl or Python) have not been written explicitly to express <code>simmap</code> data, those parsers will nonetheless be able to successfully parse this file and expose the <code>simmap</code> data to the user.</p>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bollback_2006" class="csl-entry">
Bollback, JonathanP. 2006. <em>BMC Bioinformatics</em> 7 (1): 88. <a href="https://doi.org/10.1186/1471-2105-7-88" class="external-link">https://doi.org/10.1186/1471-2105-7-88</a>.
</div>
<div id="ref-Huelsenbeck_2003" class="csl-entry">
Huelsenbeck, John P., Rasmus Nielsen, and Jonathan P. Bollback. 2003. <span>“Stochastic Mapping of Morphological Characters.”</span> <em>Systematic Biology</em> 52 (2): 131–58. <a href="https://doi.org/10.1080/10635150390192780" class="external-link">https://doi.org/10.1080/10635150390192780</a>.
</div>
<div id="ref-Revell_2012" class="csl-entry">
Revell, Liam J. 2012. <span>“Phytools: An r Package for Phylogenetic Comparative Biology (and Other Things).”</span> <em>Methods in Ecology and Evolution</em> 3: 217–23.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><!-- begin footer --><div class="footer">
    <div class="container">
        <div class="row start top-4 bottom-8">
            <div class="col-2"> <img id="footerlogo" src="https://ropensci.org/img/icon_short_white.svg">
</div>
            <div class="col-10">
                <div class="row">
                    <div class="col-md-4 col-xs-6">
                        <a href="https://github.com/ropensci" target="_blank" class="external-link"><div class="icon fab fa-github"></div></a>
                        <a href="https://github.com/ropenscilabs" target="_blank" class="external-link"><div class="icon fa fa-flask"></div></a>
                        <a href="https://hachyderm.io/@ropensci" target="_blank" class="external-link"><div class="icon fab fa-mastodon"></div></a>
                        <a href="https://vimeo.com/ropensci" target="_blank" class="external-link"><div class="icon fab fa-vimeo"></div></a>
                    </div>
                </div>
                <div class="row top-4">
                    <div class="col-md-2 col-sm-4">
                        <ul>
<h5 class="bottom-2">About</h5>
                            <li><a href="https://ropensci.org/about" class="external-link">About rOpenSci</a></li>
                            <li><a href="https://ropensci.org/software-review" class="external-link">Software Review</a></li>
                            <li><a href="https://ropensci.org/about#team" class="external-link">Our Team</a></li>
                            <li><a href="https://ropensci.org/careers" class="external-link">Jobs</a></li>
                            <li><a href="https://ropensci.org/donate" class="external-link">Donate</a></li>
                            <li><a href="https://ropensci.org/contact" class="external-link">Contact Us</a></li>
                        </ul>
</div>
                    <div class="col-md-3 col-sm-4">
                        <ul>
<h5 class="bottom-2">Community</h5>
                            <li><a href="https://ropensci.org/community/" class="external-link">Our Community</a></li>
                            <li><a href="https://ropensci.org/commcalls/" class="external-link">Community calls</a></li>
                            <li><a href="https://ropensci.org/events/" class="external-link">Events</a></li>
                            <li><a href="https://discuss.ropensci.org/" class="external-link">Join the Discussion</a></li>
                            <li><a href="https://ropensci.org/code-of-conduct" class="external-link">Code of conduct</a></li>
                        </ul>
</div>
                    <div class="col-md-2 col-sm-4">
                        <ul>
<h5 class="bottom-2">Resources</h5>
                            <li><a href="https://ropensci.org/packages/" class="external-link">Packages</a></li>
                            <li><a href="https://ropensci.org/usecases/" class="external-link">Use Cases</a></li>
                            <li><a href="https://ropensci.org/talks-papers/" class="external-link">Talks &amp; Publications</a></li>
                            <li><a href="https://docs.ropensci.org/" class="external-link">Documentation</a></li>
                            <li><a href="https://ropensci.org/news/" class="external-link">Newsletter</a></li>
                            <li><a href="https://ropensci.org/how-to-cite-ropensci/" class="external-link">Cite rOpenSci</a></li>
                        </ul>
</div>
                    <div class="col-md-4 col-xs-12">
                        <h5 class="bottom-2"></h5>

                        <p>rOpenSci is a fiscally sponsored project of <a href="http://numfocus.org" class="external-link">NumFOCUS</a>.</p>

                    </div>
                </div>
            </div>
        </div>
    </div>
</div>
<!-- / end footer -->


    </footer>
</div>

  

  

  </body>
</html>
